File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
internal class RemoteControlCar
{
	private readonly int _speed;
	private readonly int _batteryDrain;

	public RemoteControlCar(int speed, int batteryDrain)
	{
		_speed = speed;
		_batteryDrain = batteryDrain;
	}

	private int _distanceDriven = 0;
	private int _batteryLife = 100;

	public bool BatteryDrained() =>
		_batteryLife < _batteryDrain;

	public int DistanceDriven() =>
		_distanceDriven;

	public void Drive()
	{
		if (!BatteryDrained())
		{
			_distanceDriven += _speed;
			_batteryLife -= _batteryDrain;
		}
	}

	public static RemoteControlCar Nitro() =>
		new(50, 4);
}

internal class RaceTrack
{
	private readonly int _distance;

	public RaceTrack(int distance) => _distance = distance;

	public bool TryFinishTrack(RemoteControlCar car)
	{
		int distanceDriven = 0;
		while (!car.BatteryDrained() && !IsComplete(distanceDriven))
		{
			car.Drive();
			distanceDriven = car.DistanceDriven();
		}

		return IsComplete(distanceDriven);
	}

	private bool IsComplete(int distanceDriven) =>
		distanceDriven >= _distance;
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
	private final int speed;
	private final int batteryDrain;
    NeedForSpeed(int speed, int batteryDrain) {
		this.speed = speed;
		this.batteryDrain = batteryDrain;
    }

	private int battery = 100;
	private int distanceDriven = 0;

    public boolean batteryDrained() {
        return this.battery < this.batteryDrain;
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
		if (!this.batteryDrained()) {
			this.distanceDriven += this.speed;
			this.drainBattery();
		}
    }

	private void drainBattery() {
		this.battery -= batteryDrain;
		if (this.battery < 0) {
			this.battery = 0;
		}
	}

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
	private final int distance;
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
		while (!car.batteryDrained() && car.distanceDriven() < this.distance) {
			car.drive();
		}

		return this.distance <= car.distanceDriven();
    }
}
